An Old Timer's Tears

ActuallyTaylor opened the pull request:
Refactors timers to use a common core that can be subclassed by other timer like objects (PWM atm).

## Additions:
- `Timer`: A `.hpp` only class that defines what a timer looks like. Also contains the definition for `TimerConfiguration`.

## Changes:
- `TimerF4xx` is a`cpp & hpp` file that defines a `Timer` implementation for for the F3 boards.
- `TimerF3xx` is a `cpp & hpp` file that defines a `Timer` implementation for for the F4 boards.
- `PWMf3xx is now a subclass of TimerF3xx, allows for PWM to control its own timers using the standard F3 timer implementation.
- `ADC` F3 & F4: Both now use the new timer implementations.

## Tested
- [x] All timers on F302
- [x] All timers on F334
- [x] All timers on F446
- [x] PWM on F302
- [x] PWM on F334
- [x] ADC on F302
- [x] ADC on F334
- [x] ADC on F446

ActuallyTaylor commented:
I need to work on documentation... But that will come over the next few days

ActuallyTaylor commented on include/core/dev/RTCTimer.hpp:
```diff
@@ -40,7 +40,7 @@ class RTCTimer : public Timer {
     void reloadTimer() override;
 
-    void setPeriod(uint32_t clock) override;
+    void setPeriod(uint32_t clock);
```
Technically that function does not have the correct signature. The correct signature includes a prescaler, which the RTC timer does not have. 

I have updated this to just always take in the AUTO_PRESCALER and do nothing with it. Because it is definitely more confusing to not override the function.

ActuallyTaylor commented on include/core/dev/Timer.hpp:
```diff
@@ -37,10 +89,10 @@ class Timer {
      * You must call startTimer again to continue timer operation.
      *
      * @param[in] clockPeriod the clock period in ms.  An interrupt will be triggered at that frequency.
```
Why would the ms resolution only be setup when `AUTO_PRESCALER` is set?

mjh9585 replied:
Well changing the prescalar is going to affect all of your timing calculations so the units would no longer in ms but in timer counts. If your HAL_RCC_GetHCLKFreq() / 1000 with `AUTO_PRESCALER` is normally 20 then changing the prescalar to 10 would then be half milliseconds, thus your units change depending on this setting.

ActuallyTaylor commented on samples/timer/main.cpp:
```diff
@@ -43,21 +43,37 @@ int main() {
+    dev::Timer& sampleTimer3 = dev::getTimer<dev::MCUTimer::Timer12>(200);
+    #else
+    dev::Timer& sampleTimer2 = dev::getTimer<dev::MCUTimer::Timer15>(1000, configuration);
+    dev::Timer& sampleTimer3 = dev::getTimer<dev::MCUTimer::Timer16>(1000, configuration);
```
Purely forgot to add them

ActuallyTaylor commented on samples/pwm/main.cpp:
```diff
@@ -4,25 +4,29 @@
 #include <core/utils/time.hpp>
 
 namespace io   = core::io;
-namespace time = core::time;
```
It was removed for some weird build issues. I added it back and it seems to be okay.

ActuallyTaylor commented on src/core/dev/platform/f3xx/Timerf3xx.cpp:
```diff
@@ -1,32 +1,164 @@
+#define F302_TIMER_COUNT 5
+
+TIM_HandleTypeDef halTimers[F302_TIMER_COUNT];
+void (*timerInterruptHandlers[F302_TIMER_COUNT])(void* htim) = {nullptr};
```
I think this is out of scope for this PR. This was just a refactor of timers for the future. No features were actually added, only bugs were fixed.

ActuallyTaylor replied:
I have added this in my most recent commits. Talked it over with Rue, and it is definitely worth it.

ActuallyTaylor commented on include/core/dev/platform/f3xx/Timerf3xx.hpp:
```diff
@@ -22,34 +22,65 @@ class Timerf3xx : public Timer {
+     * @param[in] clockPrescaler the prescaler that the clock will use. If clockPrescaler is set to @ref AUTO_PRESCALER, this function will calculate its own prescaler value.
      */
-    explicit Timerf3xx(TIM_TypeDef* timerPeripheral, uint32_t clockPeriod);
+    explicit TimerF3xx(TIM_TypeDef* timerPeripheral, uint32_t clockPeriod, TimerConfiguration configuration, uint32_t clockPrescaler = AUTO_PRESCALER);
```
> initialize our TIM_TypeDef in the constructor

What do you mean by this?

ActuallyTaylor commented on include/core/manager.hpp:
```diff
@@ -227,10 +227,165 @@ I2C& getI2C() {
+    default:
+        return dev::MCUTimer::None;
+    }
+}
```
Okay, my info was out of date (it was my own info idk how I managed that), but since I moved PWMF3 to a timer subclass there are actually no problems with doing this directly in the class. The original problem was that PWMF3 needed an entire timer created, and that caused include loops. NOT MCUTimer, which I originally stated.

ActuallyTaylor commented on include/core/dev/platform/f3xx/Timerf3xx.hpp:
```diff
@@ -22,34 +22,65 @@ class Timerf3xx : public Timer {
+     * @param[in] clockPrescaler the prescaler that the clock will use. If clockPrescaler is set to @ref AUTO_PRESCALER, this function will calculate its own prescaler value.
      */
-    explicit Timerf3xx(TIM_TypeDef* timerPeripheral, uint32_t clockPeriod);
+    explicit TimerF3xx(TIM_TypeDef* timerPeripheral, uint32_t clockPeriod, TimerConfiguration configuration, uint32_t clockPrescaler = AUTO_PRESCALER);
```
I debated this, but I decided not to do it. My main reasoning was that TimerConfiguration variables do not change. The prescaler can be changed at runtime by using setPeriod. This would leave the configuration object out of date.

ActuallyTaylor commented on include/core/dev/platform/f3xx/Timerf3xx.hpp:
```diff
@@ -22,34 +22,65 @@ class Timerf3xx : public Timer {
+     * @param[in] clockPeriod the clock period in ms.  An interrupt will be triggered at that frequency.
+     * @param [in] clockPrescaler the prescaler used by the clock. Divides the system clock frequency to get it within an acceptable range for clocking. If set to @ref AUTO_PRESCALER, the function implementation should calculate its own prescaler.
+     */
+    void setPeriod(uint32_t clockPeriod, uint32_t clockPrescaler) override;
```
That is totally a way to do this, but there is no two way binding back to the configuration object. I think its still better to explicitly ask for it. Especially when `AUTO_PRESCALER` exists, and is already a default argument everywhere it can be.

ActuallyTaylor commented:
I made some general updates. I have added a new default configuration & template method that will make this a drop-in replacement for the current timers. I also added a new sample **timer_configuration** that shows how to use the configuration objects.

aclowmclaughlin commented on include/core/dev/Timer.hpp:
```diff
@@ -3,14 +3,68 @@
 
-namespace core::dev {
+/**
+ * If AUTO_PRESCALER is sent as a prescaler the timer will calculate its own Prescaler.
```
Also, is it every possible that someone could want to prescale by this max value? I assume not...

aclowmclaughlin commented on include/core/dev/platform/f3xx/Timerf3xx.hpp:
```diff
@@ -22,34 +32,76 @@ class Timerf3xx : public Timer {
      * do not occur.
```
On the line above it lists possible options- are they still accurate?

aclowmclaughlin commented on include/core/dev/platform/f3xx/Timerf3xx.hpp:
```diff
@@ -22,34 +32,76 @@ class Timerf3xx : public Timer {
+     * an acceptable range for clocking. If set to @ref AUTO_PRESCALER, the function implementation should calculate its
+     * own prescaler.
+     */
+    void setPeriod(uint32_t clockPeriod, uint32_t clockPrescaler = AUTO_PRESCALER) override;
```
Do these need their own descriptions? They should inherit the descriptions from the parent class, right?

ActuallyTaylor replied:
I think I have this figured out, using `@inheritDoc`. Clion at least grabs the correct information, but I can not speak for vscode.

aclowmclaughlin commented on include/core/dev/platform/f4xx/Timerf4xx.hpp:
```diff
@@ -10,46 +10,101 @@
+     * that triggers with the given period.  Starts the timer
+     * @param timerPeripheral[in] The timer peripheral to configure. Possible options for this board are
+     * TIM1, TIM2, TIM3, TIM4, TIM5, TIM8, TIM9, TIM10, TIM11, TIM11, TIM13, TIM14.
+     * It is up to the user to verify that resource conflicts do not occur.
```
Is there no way to programmatically ensure no conflicts?

aclowmclaughlin replied:
Understood. I'd rather give people too much freedom than not enough. We don't use that many timers so I think preventing resource overuse is pretty easy.

aclowmclaughlin commented on include/core/manager.hpp:
```diff
@@ -124,14 +124,42 @@ RTC& getRTC() {
+ * @tparam mcuTimer The microcontroller timer that the timer will run on.
+ * @param clockPeriod The clock period to run the timer at
+ * @param configuration Configuration variables for the timer.
+ * @return
```
Add return description

aclowmclaughlin commented on include/core/manager.hpp:
```diff
@@ -124,14 +124,42 @@ RTC& getRTC() {
+ *
+ * @tparam mcuTimer The microcontroller timer that the timer will run on.
+ * @param clockPeriod The clock period to run the timer at
+ * @return
```
Add return description- couldn't you just have it have a default argument instead of two functions that do the same thing?

aclowmclaughlin commented on samples/timer_configuration/main.cpp:
```diff
@@ -0,0 +1,79 @@
+    reloadGPIO             = &io::getGPIO<io::Pin::PC_0>(io::GPIO::Direction::OUTPUT);
+
+    dev::TimerConfiguration_t configuration = {
+        TIM_COUNTERMODE_UP,            // Tell the counter to count upwards towards the clock period
```
These should be enums in the Timer class that represent these, instead of the macros. This will allow them to be commented much better AND I think that I prefer any values a sort of "end user" (i.e. firmware member writing non-HAL level code) do not come from the HAL.

aclowmclaughlin commented on src/core/dev/platform/f3xx/Timerf3xx.cpp:
```diff
@@ -1,32 +1,164 @@
+TIM_HandleTypeDef halTimers[F334_TIMER_COUNT];
+void (*timerInterruptHandlers[F334_TIMER_COUNT])(void* htim) = {nullptr};
+
+// Timer 6 is technically available but does not support Capture Compare channels, so is therefore not supported.
```
Love this commenting!

aclowmclaughlin reviewed (changes requested):
This is super impressive and it is really well detailed. Everything is really minor and mostly commenting changes.
I really do agree that passing in an extre void* argument in the interrupts is not a significant code burden to implement but will be an incredible boon for the code.
HOWEVER, if there's more overhead to doing that then I'm really predicting, then by all means we should probably have it be a separate request, I'm just thinking that it won't be that big a burden.

aclowmclaughlin reviewed (approved):
This looks really good to me. Stellar work.
I had a couple of questions but I don't think any of them were important enough to stop me from approving the PR.

ActuallyTaylor commented on include/core/dev/RTCTimer.hpp:
```diff
@@ -67,13 +75,13 @@ class RTCTimer : public Timer {
 
     /** The amount of time it takes the timer to go off in SECONDS */
-    uint32_t clockPeriod;
+    uint32_t clockPeriod{};
```
I prefer { }, but I think I am in the minority on that. I will swap this over & also add in the defaults for the rest of the values.

ActuallyTaylor commented on samples/pwm/main.cpp:
```diff
@@ -2,8 +2,9 @@
  * analyzer to see the generated square wave.
  */
-#include <core/io/PWM.hpp>
+// #include <core/io/PWM.hpp>
```
it isn't needed because it is included through manager.hpp

ActuallyTaylor commented on src/core/io/platform/f4xx/ADCf4xx.cpp:
```diff
@@ -54,7 +55,15 @@ constexpr uint8_t ADC3_SLOT = 2;
+                                           dev::TimerMasterModeSelection::UPDATE,
+                                           dev::TimerMasterSlaveMode::DISABLE};
+
+dev::TimerF4xx core::io::ADCf4xx::timer8 = dev::TimerF4xx(TIM8, 1000, configuration);
```
I am not sure on this. This is what was present when for the F4 adc when i started the refactor. My best guess is that, that timer is just wired for the ADC.
